pkg/gardenlet/controller/backupbucket/reconciler.go

```diff
@@ -0,0 +1,467 @@
+			return reconcile.Result{}, fmt.Errorf("failed to remove finalizer from secret: %w", err)
+		}
+	}
+
+	if controllerutil.ContainsFinalizer(backupBucket, gardencorev1beta1.GardenerName) {
+		log.Info("Removing finalizer")
+		if err := controllerutils.RemoveFinalizers(ctx, r.GardenClient, backupBucket, gardencorev1beta1.GardenerName); err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to remove finalizer: %w", err)
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func updateBackupBucketStatusProcessing(ctx context.Context, c client.StatusClient, bb *gardencorev1beta1.BackupBucket, message string, progress int32, clock clock.Clock) error {
+	patch := client.MergeFrom(bb.DeepCopy())
+	bb.Status.LastOperation = &gardencorev1beta1.LastOperation{
+		Type:           v1beta1helper.ComputeOperationType(bb.ObjectMeta, bb.Status.LastOperation),
+		State:          gardencorev1beta1.LastOperationStateProcessing,
+		Progress:       progress,
+		Description:    message,
+		LastUpdateTime: metav1.NewTime(clock.Now()),
+	}
+	return c.Status().Patch(ctx, bb, patch)
+}
+
+func updateBackupBucketStatusError(ctx context.Context, c client.StatusClient, bb *gardencorev1beta1.BackupBucket, message string, lastError *gardencorev1beta1.LastError, clock clock.Clock) error {
+	patch := client.MergeFrom(bb.DeepCopy())
+
+	bb.Status.LastOperation = &gardencorev1beta1.LastOperation{
+		Type:           v1beta1helper.ComputeOperationType(bb.ObjectMeta, bb.Status.LastOperation),
+		State:          gardencorev1beta1.LastOperationStateError,
+		Progress:       50,
+		Description:    message,
+		LastUpdateTime: metav1.NewTime(clock.Now()),
+	}
+	bb.Status.LastError = lastError
+
+	return c.Status().Patch(ctx, bb, patch)
+}
+
+func updateBackupBucketStatusSucceeded(ctx context.Context, c client.StatusClient, bb *gardencorev1beta1.BackupBucket, message string, clock clock.Clock) error {
+	patch := client.MergeFrom(bb.DeepCopy())
+
+	bb.Status.LastError = nil
+	bb.Status.LastOperation = &gardencorev1beta1.LastOperation{
+		Type:           v1beta1helper.ComputeOperationType(bb.ObjectMeta, bb.Status.LastOperation),
+		State:          gardencorev1beta1.LastOperationStateSucceeded,
+		Progress:       100,
+		Description:    message,
+		LastUpdateTime: metav1.NewTime(clock.Now()),
+	}
+	bb.Status.ObservedGeneration = bb.Generation
+
+	return c.Status().Patch(ctx, bb, patch)
+}
+
+func (r *Reconciler) emptyExtensionSecret(backupBucketName string) *corev1.Secret {
+	return &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      generateBackupBucketSecretName(backupBucketName),
+			Namespace: r.GardenNamespace,
+		},
+	}
+}
+
+func generateBackupBucketSecretName(backupBucketName string) string {
+	return fmt.Sprintf("bucket-%s", backupBucketName)
+}
+
+func generateGeneratedBackupBucketSecretName(backupBucketName string) string {
+	return v1beta1constants.SecretPrefixGeneratedBackupBucket + backupBucketName
+}
```

Comment:
Nit: Can you group non-`*Reconciler` methods below after all `*Reconciler` methods?